qualtran/bloqs/arithmetic/sorting.py

```diff
@@ -11,118 +11,318 @@
+    r"""Compare and potentially swaps two numbers in {0, ... L - 1}.
 
     Implements $U|a\rangle|b\rangle|0\rangle \rightarrow |\min(a,b)\rangle|\max(a,b)\rangle|a>b\rangle$,
 
     where $a$ and $b$ are n-qubit quantum registers. On output a and b are
     swapped if a > b. Forms the base primitive for sorting.
+    The numbers are represented using $n = \ceil{\log L}$ bits.
 
     Args:
-        bitsize: Number of bits used to represent each integer.
+        L: Upper-bound (excluded) on the input numbers.
 
     Registers:
-        a: A nbit-sized input register (register a above).
-        b: A nbit-sized input register (register b above).
+        a: A n-bit-sized input register (register a above).
+        b: A n-bit-sized input register (register b above).
         out: A single bit output register which will store the result of the comparator.
 
     References:
         [Improved techniques for preparing eigenstates of fermionic Hamiltonians](https://www.nature.com/articles/s41534-018-0071-5).
         Fig. 1. in main text.
     """
 
-    bitsize: SymbolicInt
+    L: SymbolicInt
 
     @property
     def signature(self):
+        input_dtype = BoundedQUInt(self.bitsize, self.L)
         return Signature(
             [
-                Register('a', QAny(self.bitsize)),
-                Register('b', QAny(self.bitsize)),
+                Register('a', input_dtype),
+                Register('b', input_dtype),
                 Register('out', QBit(), side=Side.RIGHT),
             ]
         )
 
+    @property
+    def bitsize(self):
+        """number of bits to represent the inputs"""
+        return bit_length(self.L - 1)
+
     def pretty_name(self) -> str:
         return "Cmprtr"
 
-    def _t_complexity_(self):
-        # complexity is from less than on two n qubit numbers + controlled swap
-        # Hard code for now until CSwap-Bloq is merged.
-        # See: https://github.com/quantumlib/Qualtran/issues/219
-        t_complexity = GreaterThan(self.bitsize, self.bitsize).t_complexity()
-        t_complexity += TComplexity(t=14 * self.bitsize)
-        return t_complexity
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', a: 'Soquet', b: 'Soquet'
+    ) -> dict[str, 'SoquetT']:
+        out = bb.allocate(dtype=QBit())
+        a, b, out = bb.add(GreaterThan(self.bitsize, self.bitsize), a=a, b=b, target=out)
+        out, a, b = bb.add(CSwap(self.bitsize), ctrl=out, x=a, y=b)
+        return {'a': a, 'b': b, 'out': out}
 
 
 @bloq_example
 def _cmp_symb() -> Comparator:
-    n = sympy.Symbol('n')
-    cmp_symb = Comparator(bitsize=n)
+    L = sympy.Symbol('L')
+    cmp_symb = Comparator(L=L)
     return cmp_symb
 
 
 _COMPARATOR_DOC = BloqDocSpec(bloq_cls=Comparator, examples=[_cmp_symb])
 
 
 @frozen
-class BitonicSort(Bloq):
-    r"""Sort k n-bit numbers.
+class ParallelComparators(Bloq):
+    """Given k numbers in [0, L), compare every pair that is `offset` apart.
+
+    Args:
```

Comment:
Please more details to the docstring - do we compare and swap or we just store the result of the comparison somewhere? Are the RIGHT junk registers storing the result of the comparison? Or are they supposed to be completely ignored?

Reply:
updated all docstrings.